File: Practice/src/introduction_coding_tests/NumberOfWaysToDivideBeads.java
package introduction_coding_tests;

import java.math.BigInteger;

public class NumberOfWaysToDivideBeads {
//	구슬을 나누는 경우의 수
	public int solution(int balls, int share) {
		BigInteger answer = new BigInteger("1");
		for (int i = balls; i > share; i--) {
			answer = answer.multiply(new BigInteger(i + ""));
		}
		for (int i = share - 1; i > 1; i--) {
			answer = answer.divide(new BigInteger(i + ""));
		}
		return answer.intValue();
	}

//	public long solution(int balls, int share) {
//		share = Math.min(balls - share, share);
//
//		if (share == 0)
//			return 1;
//
//		long result = solution(balls - 1, share - 1);
//		result *= balls;
//		result /= share;
//
//		return result;
//	}

}

File: Practice/src/introduction_coding_tests/NumberOfWaysToDivideBeads.js
function solution(balls, share) {
    var answer = 1;
    const arr = Array(31).fill(0);
    for (let i = share + 1; i <= balls; i++) {
        let temp = i;
        for (let j = 2; temp > 1; ) {
            if (temp % j == 0) {
                arr[j]++;
                temp /= j;
            } else {
                j++;
            }
        }
    }

    for (let i = balls - share; i > 1; i--) {
        let temp = i;
        for (let j = 2; temp > 1; ) {
            if (temp % j == 0) {
                arr[j]--;
                temp /= j;
            } else {
                j++;
            }
        }
    }
    for (let i = arr.length - 1; i > 1; i--) {
        if (arr[i] > 0) answer *= Math.pow(i, arr[i]);
        else if (arr[i] < 0) answer /= Math.pow(i, -arr[i]);
    }
    return answer;
}
